# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [1]:
%pip install pandas scipy numpy

  Using cached pandas-2.3.2-cp313-cp313-macosx_11_0_arm64.whl.metadata (91 kB)
  Using cached scipy-1.16.1-cp313-cp313-macosx_14_0_arm64.whl.metadata (61 kB)
  Using cached numpy-2.3.2-cp313-cp313-macosx_14_0_arm64.whl.metadata (62 kB)
  Using cached pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)
Using cached pandas-2.3.2-cp313-cp313-macosx_11_0_arm64.whl (10.7 MB)
Using cached scipy-1.16.1-cp313-cp313-macosx_14_0_arm64.whl (20.8 MB)
Using cached numpy-2.3.2-cp313-cp313-macosx_14_0_arm64.whl (5.1 MB)
Using cached pytz-2025.2-py2.py3-none-any.whl (509 kB)
Using cached tzdata-2025.2-py2.py3-none-any.whl (347 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5/5 [pandas]2m4/5 [pandas]
Note: you may need to restart the kernel to use updated packages.


In [2]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np



In [5]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0         Bulbasaur    Grass  Poison  45      49       49       65       65   
1           Ivysaur    Grass  Poison  60      62       63       80       80   
2          Venusaur    Grass  Poison  80      82       83      100      100   
3     Mega Venusaur    Grass  Poison  80     100      123      122      120   
4        Charmander     Fire     NaN  39      52       43       60       50   
..              ...      ...     ...  ..     ...      ...      ...      ...   
795         Diancie     Rock   Fairy  50     100      150      100      150   
796    Mega Diancie     Rock   Fairy  50     160      110      160      110   
797  Hoopa Confined  Psychic   Ghost  80     110       60      150      130   
798   Hoopa Unbound  Psychic    Dark  80     160       60      170      130   
799       Volcanion     Fire   Water  80     110      120      130       90   

     Speed  Generation  Legendary  
0       45           1      False  
1       60           1      False  
2       80           1      False  
3       80           1      False  
4       65           1      False  
..     ...         ...        ...  
795     50           6       True  
796    110           6       True  
797     70           6       True  
798     80           6       True  
799     70           6       True  

[800 rows x 11 columns]

- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

In [10]:
len(df[df['Type 1'] == 'Dragon']['HP'])

32

In [11]:
# H0: Dragon mu HP >= Grass mu HP
# H1: Dragon mu HP < Grass mu HP
# We consider a sample of 40 Pokemons from each type
# Significance level = 0.05

dragon = df[df['Type 1'] == 'Dragon']['HP'].sample(30)
grass = df[df['Type 1'] == 'Grass']['HP'].sample(30)

st.ttest_ind(dragon, grass, alternative='less')

TtestResult(statistic=np.float64(2.7001863481114143), pvalue=np.float64(0.9954662680969679), df=np.float64(58.0))

We are not able to reject H0 hypothesis: no significant evidence

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


In [22]:
# H0: Legendary mu stats = Non Legendary mu stats
# H1: Legendary mu stats != Non Legendary mu stats
# We consider a sample of 40 Pokemons from each type
# Significant level = 0.05

non_legendary = df.loc[df['Legendary'] == False, ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].sample(40)

legendary = df.loc[df['Legendary'] == True, ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']].sample(40)

st.ttest_ind(non_legendary, legendary, alternative='two-sided')

TtestResult(statistic=array([-6.16755672, -6.16500997, -4.60731302, -7.41991116, -7.3658484 ,
       -5.41183674]), pvalue=array([2.87308611e-08, 2.90433066e-08, 1.56437783e-05, 1.23935302e-10,
       1.57405763e-10, 6.65564558e-07]), df=array([78., 78., 78., 78., 78., 78.]))

We reject the null hypothesis because all the p-values are lower than the significance level

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [23]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.31     34.19                15.0       5612.0          1283.0   
1    -114.47     34.40                19.0       7650.0          1901.0   
2    -114.56     33.69                17.0        720.0           174.0   
3    -114.57     33.64                14.0       1501.0           337.0   
4    -114.57     33.57                20.0       1454.0           326.0   

   population  households  median_income  median_house_value  
0      1015.0       472.0         1.4936             66900.0  
1      1129.0       463.0         1.8200             80100.0  
2       333.0       117.0         1.6509             85700.0  
3       515.0       226.0         3.1917             73400.0  
4       624.0       262.0         1.9250             65500.0

**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 34)
- Hospital coordinates (-122, 37)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [28]:
# Define the hospital coordinates
hospital_lon, hospital_lat = -122.0, 37.0

# Define the school coordinates
school_lon, school_lat = -118.0, 34.0

# Compute Euclidean distance to hospital and assign to new column
df['distance_to_hospital'] = np.sqrt(
    (df.longitude - hospital_lon)**2 +
    (df.latitude  - hospital_lat)**2
)

# Compute Euclidean distance to school and assign to new column
df['distance_to_school'] = np.sqrt(
    (df.longitude - school_lon)**2 +
    (df.latitude  - school_lat)**2
)

# Create the boolean column
df["close_to_services"] = (
    (df["distance_to_hospital"] < 0.5) |
    (df["distance_to_school"]  < 0.5)
)

df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  \
0          1015.0       472.0         1.4936             66900.0   
1          1129.0       463.0         1.8200             80100.0   
2           333.0       117.0         1.6509             85700.0   
3           515.0       226.0         3.1917             73400.0   
4           624.0       262.0         1.9250             65500.0   
...           ...         ...            ...                 ...   
16995       907.0       369.0         2.3571            111400.0   
16996      1194.0       465.0         2.5179             79000.0   
16997      1244.0       456.0         3.0313            103600.0   
16998      1298.0       478.0         1.9797             85800.0   
16999       806.0       270.0         3.0147             94600.0   

       distance_to_hospital  distance_to_school  close_to_services  
0                  8.187319            3.694888              False  
1                  7.966235            3.552591              False  
2                  8.143077            3.453940              False  
3                  8.154416            3.448840              False  
4                  8.183508            3.456848              False  
...                     ...                 ...                ...  
16995              4.233675            9.082070              False  
16996              4.332320            9.168915              False  
16997              5.358694           10.057614              False  
16998              5.322593           10.026465              False  
16999              4.249012            9.115597              False  

[17000 rows x 12 columns]

In [33]:
# H0: houses close to either a school or a hospital price >  houses far from either a school or a hospital
# H1: houses close to either a school or a hospital price <= houses far from either a school or a hospital
# Significant level = 0.05

far = df[df['close_to_services'] == False]['median_house_value'].sample(40)
close = df[df['close_to_services'] == True]['median_house_value'].sample(40)

st.ttest_ind(close, far, alternative='less')

TtestResult(statistic=np.float64(2.97754414727049), pvalue=np.float64(0.9980651161168688), df=np.float64(78.0))

we fail to reject the null hypothesis because the p-value is greater than the significant level